design/transaction-state-store.md

```diff
@@ -0,0 +1,99 @@
+# Transaction State Store (txnStateStore)
+
+This document describes the transaction state store (often is referred as `txnStateStore` in the code) in FDB. The transaction state store keeps important metadata about the database to bootstrap the database, to guide the transaction system to persist writes (i.e., help assign storage tags to mutations at commit proxies), and to manage data (i.e., shard) movement metadata. This is a critical piece of information that have to be consistent across many processes and to be persistent for recovery.
+
+Acknowledgment: A lot of contents are taken from [Evan's FDB brownbag talk](https://drive.google.com/file/d/15UvKiNc-jSFfDGygNmLQP_d4b14X3DAS/).
+
+## What information is stored in transaction state store?
+
+The information includes: shard mapping (key range to storage server mapping, i.e.,
+`keyServers`), storage server tags (`serverTags`), tagLocalityList, storage server tag
+history, database locked flag, metadata version, mustContainSystemMutations, coordinators,
+storage server interface (`serverList`), database configurations, TSS mappings and
+quarantines, backup apply mutation ranges and log ranges, etc.
+
+The information of transaction state store is kept in the system key space, i.e., using the
+`\xff` prefix. Note all data in the system key space are saved on storage servers. The
+`txnStateStore` is only a part of the `\xff` key space, and is additionally kept in the
+memory of commit proxies as well as disks of the log system (i.e., TLogs). Changes to
+the `txnStateStore` are special mutations to the `\xff` key space, and are called
+inconsistently in the code base as "metadata mutations" in commit proxies and
+"state mutations" in Resolvers.
+
+## Why do we need transaction state store?
+
+When bootstraping an FDB cluster, the new master (i.e., the sequencer) role recruits a
+new transaction system and initializes them. In particular, the transaction state store
+is first read by the master from previous generation's log system, and then broadcast to
+all commit proxies of the new transaction system. After initializing `txnStateStore`, these
+commit proxies know how to assign mutations with storage server tags: `txnStateStore`
+contains the shard map from key range to storage servers; commit proxies use the shard
+map to find and attach the destination storage tags for each mutation.
+
+## How is transaction state store replicated?
```

Comment:
When I learned about the `txnStateStore` initially it was a bit confusing to me how a client would be able to read from it. I was looking for code where the proxies would provide a `StorageServerInterface` to serve reads to keys in this storage.

To make sure nobody else runs into this confusion I would recommend you also mention that everything in the `txnStateStore` is written as a normal key range to storage servers and it is written to the `txnStateStore`. So a client (for example `fdbcli`) will read from the storage servers but then the commit will go to the `txnStateStore` AND to the storages responsible for these shards.

Reply:
Done.